pkg/gadgets/interface.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2022-2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gadgets
+
+import (
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/parser"
+)
+
+// GadgetType defines how a gadget is actually run
+type GadgetType string
+
+const (
+	TypeTrace          GadgetType = "trace"          // Normal trace gadgets
+	TypeTraceIntervals GadgetType = "traceIntervals" // top gadgets expecting arrays of events
+	TypeOneShot        GadgetType = "oneShot"        // Gadgets that only fetch results
+	TypeProfile        GadgetType = "profile"        // Gadgets that run until the user stops, or it times out and then shows results
+)
+
+func (t GadgetType) CanSort() bool {
+	return t == TypeOneShot || t == TypeTraceIntervals
+}
+
+func (t GadgetType) IsPeriodic() bool {
+	return t == TypeTraceIntervals
+}
```

Comment:
Hmm, two reasons mainly why I think those should stay:

* (category) names can change 🤣 and those types are bound to "how the gadget works"
* there might be more than one category that for each type (audit/seccomp is for example of TypeTrace right now)

Now what they're actually used for:

* enable certain functionality prior to running the gadgets and without checking their interfaces and doing some kind of estimation from that (we currently enable sorting/filters and so on based on the type)
* control whole UIs prior to running gadgets: based on the type we know whether we expect a flow of data and can show a table or a blob of data that we might be able to select an output format before running

IMHO, we should have as few of these types as necessary while being able to have as many categories as needed.

Reply:
> (category) names can change rofl 

W can handle it by changing the value of a constant in the code base. I don't expect to support gadgets as plugins anytime soon, so it doesn't seem to be a problem.

> "how the gadget works"

That's what we're trying to achieve by having categories. All gadgets in the same category should behave the same. 

> there might be more than one category that for each type (audit/seccomp is for example of TypeTrace right now)

We can handle all categories in the different switch statements we have around, right? 

> IMHO, we should have as few of these types as necessary while being able to have as many categories as needed.

I'm not totally sure as I think it adds complexity and confusion. Categories are there to group gadgets that behave in a similar way. It's weird to have a second field describing how the gadget runs. 

I understand that we could have something extra to describe the "capabilities" of a gadget (like a bitfield), however I think we can avoid doing that right now.

What do other folks think? /cc @alban @blanquicet

Reply:
> > (category) names can change rofl
> 
> W can handle it by changing the value of a constant in the code base. I don't expect to support gadgets as plugins anytime soon, so it doesn't seem to be a problem.
> 
> > "how the gadget works"
> 
> That's what we're trying to achieve by having categories. All gadgets in the same category should behave the same.
> 
> > there might be more than one category that for each type (audit/seccomp is for example of TypeTrace right now)
> 
> We can handle all categories in the different switch statements we have around, right?
> 
> > IMHO, we should have as few of these types as necessary while being able to have as many categories as needed.
> 
> I'm not totally sure as I think it adds complexity and confusion. Categories are there to group gadgets that behave in a similar way. It's weird to have a second field describing how the gadget runs.

Sure, for us it's easy to switch names - but for third-party users (and alternative frontends) this would mean that their parts would have to be updated as well to maintain compatibility, thus I think having an attribute that explicitly indicates what a gadget does, can in the end simplify things rather than complicate.

Reply:
> What do you think about using a bitmask instead?

Hmm, what would the bits represent? Right now, I don't see too many options/features that aren't mutually exclusive.

* use intervals
* return events instead of a "result"
* ...

However, this might complicate things more than those "fixed" types (especially if the list grows), as - theoretically - you'd have to think about every possible combination.

Reply:
Yes, that's not so easy to define. I thought about something like:

- OutputArray: The gadget outputs data in an array format.
- StartStop: The gadget that needs to be started and stopped. (Opposite to OneShot). It can produce output from the moment it's started or only  when stopped.
- Stream :  The gadget that streams events.

Given that, we can define the current gadgets as:

-  Tracers = StartStop | Stream
- Top = OutputArray | StartStop | Stream
- Snapshot = OutputArray 
- Profile = StartStop 
- audit seccomp: StartStop | Stream
- advise seccomp-profile: StartStop ?


However I'd like to get others' opinion on this discussion.

Reply:
I understand Mauro's concern about introducing this concept, but I think they indeed bring an advantage to the 3rd party users to know how they must treat each gadget.